### PR TITLE
Email Service

### PR DESCRIPTION
**Pull request by rathoreprem**
1. Create an Email Service, Service receive email with XML as a attachment, do coding for fetching XML document with it.
2. If Loan status changed , send mail to applicant, is it approved or reject. (Create a new Custom Object
Loan with Applicant Name, Applicant Email,Manager,Amount……).
3. When Opportunity status changed send mail to Owner using Email Template by Apex Code.

### CODE REVIEW

File: Email Service/EmailService.cls [hunk 1]
```diff
@@ -0,0 +1,72 @@
+global class EmailService implements Messaging.InboundEmailHandler {
```
Comment (meta-vijay-khatri) [inline Email Service/EmailService.cls:1]: why is this class global?

Comment (meta-vijay-khatri) [inline Email Service/EmailService.cls:1]: class name `EmailService` is too generic 
as this handleInboundEmail is for creating new contacts then we may name it as 
`CreateContactEmailService` or `ContactEmailService`

File: Email Service/EmailService.cls [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    global Messaging.InboundEmailResult handleInboundEmail(Messaging.inboundEmail email, Messaging.InboundEnvelope env){
+        Messaging.InboundEmailResult result = new Messaging.InboundEmailResult();
+        try{
+            if(email.textAttachments.size() > 0 && email.textAttachments[0].filename.endsWith('.xml'))
+            {
+                String myPlainText= '';
+                myPlainText = email.plainTextBody;
+                List<Contact> contactList = new List<Contact>();
+                if(email.textAttachments != null)
```
Comment (meta-vijay-khatri) [inline Email Service/EmailService.cls:10]: is this sensible? 
on line 5 we are checking email.textAttachments.size() and email.textAttachments[0].filename.endsWith('.xml')
I don't think email.textAttachments will be null in any case here

File: Email Service/EmailService.cls [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                {
+                    // Save attachments, if any
+                    for (Messaging.Inboundemail.TextAttachment tAttachment : email.textAttachments) {
+                        String xmlString = Blob.valueOf(tAttachment.body).toString();
+                        System.debug(xmlString);
```
Comment (meta-vijay-khatri) [inline Email Service/EmailService.cls:15]: System.debug 
again

File: Email Service/EmailService.cls [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                result.success = true;
+            }
+        }
+        catch(Exception e)
+        {
+            e.getMessage();
+        }
+        return result;
+    }
+    public static List<Contact> getContactList(String xmlString)
```
Comment (meta-vijay-khatri) [inline Email Service/EmailService.cls:32]: why is this method public ?

**Comment by meta-vijay-khatri** [comment]
Add the test class for EmailService

File: Email Service/UpdateOppotunityTrigger.trigger [hunk 5]
```diff
@@ -0,0 +1,15 @@
+trigger UpdateOppotunityTrigger on Opportunity (after update) {
```
Comment (meta-vijay-khatri) [inline Email Service/UpdateOppotunityTrigger.trigger:1]: incorrect name

File: Email Service/EmailSender.cls [hunk 6]
```diff
@@ -0,0 +1,23 @@
+public class EmailSender {
+    public static void sendOpportunityUpdateMail(List<Opportunity> emailopportunityList)
+    {
+        List<Messaging.SingleEmailMessage> allmsg = new List<Messaging.SingleEmailMessage>();
+        EmailTemplate emailTemplateRecord = [SELECT Id FROM EmailTemplate WHERE DeveloperName LIKE 'OpportunityUpdateEmail' LIMIT 1];
+        List<Opportunity> opportunityList = [SELECT Id,Name,Owner.Email FROM Opportunity Where Id in:emailopportunityList];
```
Comment (meta-vijay-khatri) [inline Email Service/EmailSender.cls:6]: It seems like you only need the opportunity Ids here so it's better if this method accept List<Id>

File: Email Service/EmailSender.cls [hunk 7]
```diff
@@ -0,0 +1,23 @@
+public class EmailSender {
+    public static void sendOpportunityUpdateMail(List<Opportunity> emailopportunityList)
+    {
+        List<Messaging.SingleEmailMessage> allmsg = new List<Messaging.SingleEmailMessage>();
+        EmailTemplate emailTemplateRecord = [SELECT Id FROM EmailTemplate WHERE DeveloperName LIKE 'OpportunityUpdateEmail' LIMIT 1];
```
Comment (meta-vijay-khatri) [inline Email Service/EmailSender.cls:5]: you should include EmailTemplate metadata also in the PR

File: Email Service/EmailSender.cls [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        List<Opportunity> opportunityList = [SELECT Id,Name,Owner.Email FROM Opportunity Where Id in:emailopportunityList];
+        for(Opportunity opportunityRecord: opportunityList)
+        {
+            Messaging.SingleEmailMessage mail = Messaging.renderStoredEmailTemplate(emailTemplateRecord.Id, opportunityRecord.OwnerId, opportunityRecord.Id);
+            mail.setToAddresses(new List<String>{opportunityRecord.Owner.Email});
+            System.debug(opportunityRecord.Owner.Email);
```
Comment (meta-vijay-khatri) [inline Email Service/EmailSender.cls:11]: System.debug again

**Comment by meta-vijay-khatri** [comment]
Missing 
- emailTemplate metadata
- loan object metadata
- loan status change exercise

**Review by meta-vijay-khatri — CHANGES_REQUESTED** [review]

File: Email Service/UpdateOppotunityTrigger.trigger (hunk 5 above)
Comment (rathoreprem) [reply Email Service/UpdateOppotunityTrigger.trigger:1]: changed name

File: Email Service/EmailSender.cls (hunk 6 above)
Comment (rathoreprem) [reply Email Service/EmailSender.cls:6]: ok

File: Email Service/EmailSender.cls (hunk 8 above)
Comment (rathoreprem) [reply Email Service/EmailSender.cls:11]: deleted it.

File: Email Service/EmailService.cls (hunk 4 above)
Comment (rathoreprem) [reply Email Service/EmailService.cls:32]: I remove it.

File: Email Service/EmailService.cls (hunk 3 above)
Comment (rathoreprem) [reply Email Service/EmailService.cls:15]: remove it.

File: Email Service/EmailService.cls (hunk 2 above)
Comment (rathoreprem) [reply Email Service/EmailService.cls:10]: yes, I changed it.

File: Email Service/EmailService.cls (hunk 1 above)
Comment (rathoreprem) [reply Email Service/EmailService.cls:1]: changed it.

**Pull request merged**